dimod/serialization/fileview.py

```diff
@@ -12,15 +12,109 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 #
-# =============================================================================
+"""
+A format for saving large binary quadratic models.
+
+This format is inspired by the `NPY format`_
+
+.. _NPY format: https://docs.scipy.org/doc/numpy/reference/generated/numpy.lib.format.html
+
+Extension Convention
+--------------------
+
+Binary quadratic models are typically saved using the `.bqm` extension.
+
+
+Format Version 1.0
+------------------
+
+Format specification:
+
+The first 8 bytes are a magic string: exactly "DIMODBQM".
+
+The next 1 byte is an unsigned byte: the major version of the file format.
+
+The next 1 byte is an unsigned byte: the minor version of the file format.
+
+The next 4 bytes form a little-endian unsigned int, the length of the header
+data HEADER_LEN.
+
+The next HEADER_LEN bytes form the header data. This is a json-serialized
+dictionary. The dictionary is exactly:
+
+.. code-block:: python
+
+    dict(shape=bqm.shape,
+         dtype=bqm.dtype.name,
+         itype=bqm.itype.name,
+         ntype=bqm.ntype.name,
+         vartype=bqm.vartype.name,
+         type=type(bqm).__name__,
+         variables=list(bqm.variables),
+         )
+
+it is terminated by a newline character and padded with spaces to make the
+entire length of the entire header divisible by 16.
+
+The binary quadratic model data comes after the header. The number of bytes
+can be determined by the data types and the number of variables and number
+of interactions (described in the `shape`).
+
+The first `dtype.itemsize` bytes are the offset.
+The next `num_variables * (ntype.itemsize + dtype.itemsize) bytes are the
+linear data. The linear data includes the neighborhood starts and the biases.
+The final `2 * num_interactions * (itype.itemsize + dtype.itemsize) bytes
+are the quadratic data. Stored as `(outvar, bias)` pairs.
+
+Format Version 2.0
+------------------
+
+In order to make the header a more reasonable length, the variable labels
+have been moved to the body. The `variables` field of the header dictionary
+now has a boolean value, making the dictionary:
+
+.. code-block:: python
+
+    dict(shape=bqm.shape,
+         dtype=bqm.dtype.name,
+         itype=bqm.itype.name,
+         ntype=bqm.ntype.name,
+         vartype=bqm.vartype.name,
+         type=type(bqm).__name__,
+         variables=all(v == i for i, v in enumerate(bqm.variables)),
+         )
```

Comment:
In data serialization, some redundancy is good, IMO.

Instead of requiring manual calculation of bqm data section length, I would include that in the header. Same for the variables section.

Alternatively, I would include section length as the first `uint4` of each section, as long as the section order is well defined. (One option is to also have offsets of important sections in header, but we're at only two sections, so that's not yet needed.)

Reply:
Hmm, I had tried to avoid implementing https://github.com/dwavesystems/dimod/issues/660 in the interests of not changing the format too much. But I think you're right and we should have each local header (`[OFFSET]`, `[LINEAR]`, `[QUADRATIC]`, `[VARIABLES]`) with its length. The order is well defined.

Reply:
If we're going to 2.0, might as well. Also changing the encoding from `ascii` to `utf8` a la https://numpy.org/doc/stable/reference/generated/numpy.lib.format.html#format-version-3-0

Reply:
Section headers yes, section magic no, IMO.

You don't need UTF8 because you encode user text (labels) with JSON.

Reply:
What would the header be then? Just 4 bytes of length?

Reply:
On second thought, there might be many sections, so it makes sense to identify the section type in the header. One byte would be sufficient for this, but if we want to make the file human-semi-readable, a text label works better. I'd prefer one byte for section type field, though :slightly_smiling_face:

Reply:
I think I am leaning towards having the section offsets in the header, but will try some things.

Reply:
I ended up going with 4 bytes. Seemed like a reasonable compromise between short and human readable.